torch/distributed/fsdp/common_utils.py

```diff
@@ -0,0 +1,35 @@
+"""
+Public common utilities for FSDP.
+"""
+
+from enum import auto, Enum
+
+
+class BackwardPrefetch(Enum):
+    """
+    This configures explicit backward prefetching. For NCCL backend, any
+    collectives, even if issued in different streams, contend for the same
+    per-device NCCL stream, which is why the relative order in which the
+    collectives are issued matters for overlapping. The different backward
+    prefetching settings correspond to different orderings.
+
+    - ``BACKWARD_PRE``: This prefetches the next set of parameters before the
+    current set of parameter's gradient computation. This improves backward
+    pass throughput by overlapping communication (next all-gather) and
+    computation (current gradient computation) but may increase the peak memory
+    usage since two sets of parameters are in memory at once.
+    - ``BACKWARD_POST``: This prefetches the next set of parameters after the
+    current set of parameter's gradient computation. This may improve backward
+    pass throughput by overlapping communication (current reduce-scatter) and
+    computation (next gradient computation) and does not increase the peak
+    memory usage since the current set of parameters are freed before
+    prefetching the next set. Specifically, the next all-gather is reordered to
```

Comment:
would I be correct, if we are not doing prefetch, we should have even lower memory footprint, because reduce-scatter would have deleted full grad before allgather?

Reply:
This is a good question, and it has made me take a step back about the `BACKWARD_PRE` docs too.

TL;DR: I think we should simplify the docs to say that both prefetching modes may increase peak memory and omit the explanation since it might be too confusing.

> if we are not doing prefetch, we should have even lower memory footprint, because reduce-scatter would have deleted full grad before allgather?

I am not sure if this is true. Even without prefetching, if FSDP freed the unsharded gradient, the next all-gather cannot reuse that block since it is from a different stream. Whether the next gradient computation can reuse the memory from the unsharded gradient depends on the CPU thread timing.